From 85650d0b5f2334ead139c1ff047f7860ba202f53 Mon Sep 17 00:00:00 2001
Message-Id: <85650d0b5f2334ead139c1ff047f7860ba202f53.1421709247.git.chang-joon.lee@intel.com>
In-Reply-To: <3650eb44734ad4fbe098313349da8185595a20d8.1421709247.git.chang-joon.lee@intel.com>
References: <3650eb44734ad4fbe098313349da8185595a20d8.1421709247.git.chang-joon.lee@intel.com>
From: Pallavi <pallavi.g@intel.com>
Date: Fri, 7 Nov 2014 12:00:39 +0530
Subject: [PATCH 70/75] MUST_REBASE [VPG]: adf/intel/vlv:add src and dst rec
 check

Added the check to make sure the src co-ordinates are inside the
input buffer boundary and also the dst co-ordinates
are not overflowing from the provided input range.

Added the error check to make sure the src rectangle is in
16.16 fixed point format.

MUST_REBASE: Since Google ADF framework is not in upstream and also
we cannot have two display drivers, other one being i915, we need to
work on i915 and ADF convergence path before it can be upstreamed

For: GMINL-3547
Change-Id: Ic8fc64705c1fbbff38262be0b4cc4b6895f790f8
Signed-off-by: Pallavi <pallavi.g@intel.com>
---
 drivers/video/adf/intel/core/vlv/vlv_sp_plane.c |   31 +++++++++++++++++++++++
 1 file changed, 31 insertions(+)

diff --git a/drivers/video/adf/intel/core/vlv/vlv_sp_plane.c b/drivers/video/adf/intel/core/vlv/vlv_sp_plane.c
index 8e82593..7ef7c39 100644
--- a/drivers/video/adf/intel/core/vlv/vlv_sp_plane.c
+++ b/drivers/video/adf/intel/core/vlv/vlv_sp_plane.c
@@ -392,6 +392,7 @@ static int vlv_sp_validate(struct intel_plane *plane, struct intel_buffer *buf,
 	u32 format_config, bpp;
 	bool visible = false;
 	struct rectangle clip;
+	u32 width, height;
 
 	struct rectangle src = {
 		/* sample coordinates in 16.16 fixed point */
@@ -409,6 +410,15 @@ static int vlv_sp_validate(struct intel_plane *plane, struct intel_buffer *buf,
 		.y2 = config->dst_y + config->dst_h,
 	};
 
+	/* make sure the src rectangle in 16.16 fixed point format */
+	if (!(config->src_x / (1<<16)) ||
+	    !(config->src_w / (1<<16)) ||
+	    !(config->src_y / (1<<16)) ||
+	    !(config->src_h / (1<<16))) {
+		pr_err("ADF:src rec are not in 16.16 fixed fmt%s\n", __func__);
+		return -ERANGE;
+	}
+
 	if (config->pipe->type == INTEL_PIPE_DSI) {
 		dsi_pipe = to_dsi_pipe(config->pipe);
 		mode = &dsi_pipe->config.perferred_mode;
@@ -427,6 +437,27 @@ static int vlv_sp_validate(struct intel_plane *plane, struct intel_buffer *buf,
 		return -EINVAL;
 	}
 
+	width = buf->w << 16;
+	height = buf->h << 16;
+
+	/* make sure src co-ordinates are inside the input buffer size */
+	if (config->src_w > width ||
+	    config->src_x > width - config->src_w ||
+	    config->src_h > height ||
+	    config->src_y > height - config->src_h) {
+		pr_err("ADF: Invalid source co-ordinates %s\n", __func__);
+		return -ENOSPC;
+	}
+
+	/* check against integer overflows */
+	if (config->dst_w > INT_MAX ||
+	    config->dst_x > INT_MAX - (int32_t) config->dst_w ||
+	    config->dst_h > INT_MAX ||
+	    config->dst_y > INT_MAX - (int32_t) config->dst_h) {
+		pr_err("ADF: Invalid dst co-ordinates %s\n", __func__);
+		return -ERANGE;
+	}
+
 	/* check buf limits */
 	if (buf->w < 3 || buf->h < 3 || buf->stride > 16384) {
 		pr_err("ADF: Unsutable fb for the plane %s\n", __func__);
-- 
1.7.9.5

